### DISCUSSION TITLE
How to find + rotate a block of dot-matrix text

### DISCUSSION THREAD

**Discussion by baesopa** [ImageMagick/ImageMagick#6258] [Help]
### ImageMagick version

7.1.0-Q16

### Operating system, version and so on

Windows 10 Pro

### Description

Hello, I'm an image-processing and -magick n00b; using ImageMagick-7.1.0-Q16; Windows 10.

I'm trying to set up a multi-step image composition pipeline, one of whose steps is to find and rotate a block of text. I'm hoping somebody can suggest how :)

I have several up-close images of dot-matrix text to process. Each image comtains a "block" of 2 lines of white-ish dot-matrix text (on a grey-ish background) which is the "central" part of the picture, and which is rotated arbitrarily.

I'm looking for a way to rotate the image about its center so that the text block is horizontal.

(Ideally the text would also be the right way up, but for now it's ok if it's not and I can try address in a subsequent step)

It's not necessary to re-size the image to preserve whatever would be cropped by the rotation (whatever's lost would be non-text parts of the image, and that's ok).

FYI, while original images are ~ 1.6M JPG, I resize them to 4% of the original size and save as ~ 55K PNG.

Here's a few samples of the original images.

Tia for any help!
![IMG_2021-01-25_14-48-07](https://user-images.githubusercontent.com/130934176/232341695-64fef9e0-2a97-4947-a103-e22bf4c9ce92.jpg)
![IMG_2021-01-25_14-48-09](https://user-images.githubusercontent.com/130934176/232341697-89ca0b22-df0e-465c-b906-89a6787b2238.jpg)
![IMG_2021-01-25_14-48-11](https://user-images.githubusercontent.com/130934176/232341698-c132e649-c21a-4f3c-8598-8df8dfb7682e.jpg)
![IMG_2021-01-25_14-48-19](https://user-images.githubusercontent.com/130934176/232341699-bd37efa9-b802-4c8c-8fd6-382287ae9812.jpg)

**Comment by snibgo** [comment]
> ... rotate the image about its center so that the text block is horizontal.

How precise does the result need to be?

The images have different x- and y-offsets. I suggest finding those first, so the images can be centred at the centre of the circles. Stretch the pixel values to get roughly white dots on a black background. Then search for the white spot that is furthest from the centre. In the samples you show, this will be at the extreme end of the text, the top or bottom row of dots. With that knowledge, we know the approximate angle to make the rows of dots horizontal.

**Reply by baesopa** [reply]
Thank you for the insights.

Result could be +/- 10 degrees of horizontal.

- " I suggest finding those first" : finding and centering on, say, the blue circle would be great; heck if I could find that circle I'd crop the image down to that size, which would be even better; except I don't know how to find that [partial] circle and its geometry. (There's talk of edge detection, and hough transforms, and -connected-components for concentric circles); short of rtm, could you suggest a specific approach for detecting the circle and its center?

- "search for the white spot that is furthest from the centre" ; my layman's method would be examine squares of 5x5 pixels in rol/col fashion; seems awfully rudimentary; what's a better way?

I'd hoped that one could apply a "blur" to the text so the letters bleed into one another, then trace the edges of that blob to get a quadrilateral, more or less, and get metadata on its vertices

**Comment by snibgo** [comment]
For example, Windows BAT syntax, where `n6h.jpg` is your first image:
```
magick ^
  n6h.jpg ^
  -level 55,80%% ^
  -background Blue ^
  -virtual-pixel Background ^
  -distort depolar -1,0,1840,1491 ^
  -crop 1x2@ -delete 1 +repage ^
  -flip ^
  x.jpg
```

![x](https://user-images.githubusercontent.com/33812368/232353158-7ca29856-fb2f-4f72-90cb-7e048c37d1cb.jpg)

The `-level` makes the dots roughly white on a roughly black background.

I think of "`-distort depolar -1,0,1840,1491`" as an unrolling operation. Imagine a rubber circle, and cut it along a radius from the top to the centre. Stretch it out so the sides of that radius become the sides of a rectangular image. The x-axis of that image represent an angle around the original centre ("theta"). The y-axis, from the top down, is the distance out from the centre ("rho").

"`1840,1491`" is my guess at the centre of the large circles. If my guess was exactly correct, the circles would be transformed to exactly horizontal lines.

The next stage is to find the bottom-most blue-ish circle. Crop off the bottom half of the image, and `-flip` which does a mirror top to bottom. Now we are searching for the top-most blue-ish circle.

**Reply by baesopa** [reply]
(Didn't notice this come in while I was writing above; this is def above my pay-grade and I'll need to read/reference manual/experiment to say anything even remotely productive; thx)

**Comment by baesopa** [comment]
Just to close the loop: I didn't get anywhere w/ my own attempts re the op, the "depolar" method sketched out above, or a couple others as well. As such I went the route of using commercial image processing web services -- yielded rotation angle, and more.
